react/features/vad-reporter/TrackVADEmitter.js

```diff
@@ -0,0 +1,281 @@
+import logger from './logger';
+
+/**
+ * The structure used by TrackVADEmitter to relay a score
+ */
+export type VADScore = {
+
+    /**
+     * Epoch time at which PCM was recorded
+     */
+    timestamp: number,
+
+    /**
+     * The PCM score from 0 - 1 i.e. 0.60
+     */
+    score: number,
+
+    /**
+     * Device ID associated with the VAD score
+     */
+    deviceId: string
+};
+
+/**
+ * Connects an audio JitsiLocalTrack to a RnnoiseProcessor using WebAudio ScriptProcessorNode.
+ * Once an object is created audio from the local track flows through the ScriptProcessorNode as raw PCM.
+ * The PCM is processed by the rnnoise module and a VAD (voice activity detection) score is obtained, the
+ * score is published to consumers via a callback.
+ * After work is done with this service the destroy method needs to be called for a proper cleanup.
+ */
+export default class TrackVADEmitter {
+    /**
+     * Something.
+     */
+    _micDeviceId: string;
+
+    /**
+     * The AudioContext instance.
+     */
+    _audioContext: AudioContext;
+
+    /**
+     * The JitsiLocalTrack instance.
+     */
+    _jitsiLocalTrack: Object;
+
+    /**
+     * The MediaStreamAudioSourceNode instance.
+     */
+    _audioSource: MediaStreamAudioSourceNode;
+
+    /**
+     * The ScriptProcessorNode instance.
+     */
+    _audioProcessingNode: ScriptProcessorNode;
+
+    /**
+     * Function type definition.
+     */
+    _onAudioProcess: (audioEvent: Object) => void;
+
+    /**
+     * Rnnoise adapter that allows us to calculate VAD score for PCM samples
+     */
+    _rnnoiseProcessor: RnnoiseProcessor;
+
+    /**
+     * Sample rate of the ScriptProcessorNode.
+     */
+    _procNodeSampleRate: number;
+
+    /**
+     * Buffer to hold residue PCM resulting after a ScriptProcessorNode callback
+     */
+    _bufferResidue: Float32Array;
+
+    /**
+     * Callback function with which the emitter publishes its periodic VAD scoring
+     */
+    _publishVAD: (score: VADScore) => void;
+
+    /**
+     * State flag, check if the instance was destroyed
+     */
+    _destroyed: boolean = false;
+
+    /**
+     * Constructor.
+     *
+     * @param {number} procNodeSampleRate - Sample rate of the proc node.
+     * @param {Function} publishVAD - VAD score callback function.
+     * @param {RnnoiseProcessor} rnnoiseProcessor - Rnnoise adapter that allows us to calculate VAD score
+     * for PCM samples.
+     * @param {Object} jitsiLocalTrack - JitsiLocalTrack corresponding to micDeviceId.
+     */
+    constructor(
+            procNodeSampleRate: number,
+            publishVAD: Function,
+            rnnoiseProcessor: RnnoiseProcessor,
+            jitsiLocalTrack: Object
+    ) {
+        this._procNodeSampleRate = procNodeSampleRate;
+        this._publishVAD = publishVAD;
+        this._rnnoiseProcessor = rnnoiseProcessor;
+        this._jitsiLocalTrack = jitsiLocalTrack;
+        this._micDeviceId = jitsiLocalTrack.getDeviceId();
```

Comment:
The device changed would be handled by an external component. TrackVADEmitter has no knowledge of whether or not this devices is currently being used by jitsi meet. However if the device is unplugged while a TrackVADEmitter is monitoring it, it will just return 0 VAD score on it like the device is disabled, the same service that handles the input change would have to monitor for device removed or added and take appropriate action.
As for caching, it could make sense from a RnnoiseProcessor perspective so we wouldn't have to always create a new instance, even though the cost is negligible. The jitisLocalTrack however, seems to always have a handle on the microphone when in use, and i didn't see an option to pause and resume it, which means that if we somehow stop the processing and keep it in a stale state the jitsiLocalTrack would still be active. For normal microphones that wouldn't necessarily be a problem however for bluetooth headsets this would means always having the headset mode on ( mono low quality audio ) even when the mic is not selected thus we would have to renew the localTrack when we pull it from the cache.
Considering how inexpensive creating the rnnoise instance and the above mentioned, the extra state that we would need to add for the object to be cacheable is not worth it from my perspective, but that is open for debate of-course.

Reply:
I see. Well we do use the value in several places, it's shorter and saves a couple of function calls, if you feel strongly about it I can change it.

Reply:
Nah, I see now. Leave it.